File: _site/solutions/938.range-sum-of-bst.rb
# require './aux.rb'
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end
# @param {TreeNode} root
# @param {Integer} l
# @param {Integer} r
# @return {Integer}
def range_sum_bst(root, lv, rv)
  ans = 0
  helper = lambda do |r|
    return if r.nil?
    if r.val < lv
      helper.call(r.right)
    elsif r.val > rv
      helper.call(r.left)
    else
      ans += r.val
      helper.call(r.left)
      helper.call(r.right)
    end
  end
  helper.call(root)
  ans
end
# [938] Range Sum of BST
# https://leetcode.com/problems/range-sum-of-bst/description/
# * algorithms
# * Medium (82.91%)
# * Total Accepted:    7K
# * Total Submissions: 8.4K
# * Testcase Example:  '[10,5,15,3,7,null,18]\n7\n15'
# Given the root node of a binary search tree, return the sum of values of all nodes with value between L and R (inclusive).
# The binary search tree is guaranteed to have unique values.
# Example 1:
# Input: root = [10,5,15,3,7,null,18], L = 7, R = 15
# Output: 32
# Example 2:
# Input: root = [10,5,15,3,7,13,18,1,null,6], L = 6, R = 10
# Output: 23
# Note:
#   The number of nodes in the tree is at most 10000.
#   The final answer is guaranteed to be less than 2^31.

File: python_solutions/938.range-sum-of-bst.py
from collections import Counter, defaultdict, OrderedDict, deque
from bisect import bisect_left, bisect_right 
from functools import reduce, lru_cache 
import itertools 
import math 
import string
true = True
false = False
MIN, MAX = -0x3f3f3f3f, 0x3f3f3f3f
#
# @lc app=leetcode id=938 lang=python3
#
# [938] Range Sum of BST
#
# https://leetcode.com/problems/range-sum-of-bst/description/
#
# algorithms
# Easy (79.26%)
# Total Accepted:    152K
# Total Submissions: 191.8K
# Testcase Example:  '[10,5,15,3,7,null,18]\n7\n15'
#
# Given the root node of a binary search tree, return the sum of values of all
# nodes with value between L and R (inclusive).
# 
# The binary search tree is guaranteed to have unique values.
# 
# 
# 
# 
# Example 1:
# 
# 
# Input: root = [10,5,15,3,7,null,18], L = 7, R = 15
# Output: 32
# 
# 
# 
# Example 2:
# 
# 
# Input: root = [10,5,15,3,7,13,18,1,null,6], L = 6, R = 10
# Output: 23
# 
# 
# 
# 
# Note:
# 
# 
# The number of nodes in the tree is at most 10000.
# The final answer is guaranteed to be less than 2^31.
# 
# 
# 
#
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def rangeSumBST(self, root: TreeNode, L: int, R: int) -> int:
        if not root: return 0 
        _sum = 0 
        if L <= root.val <= R: _sum += root.val 
        if L < root.val: _sum += self.rangeSumBST(root.left, L, R)
        if R > root.val: _sum += self.rangeSumBST(root.right, L, R)
        return _sum 
        

sol = Solution()


